language/move-analyzer/src/symbols.rs

```diff
@@ -0,0 +1,2184 @@
+                                    continue;
+                                }
+                            };
+                        }
+                    };
+
+                    structs.insert(
+                        *name,
+                        StructDef {
+                            name_start,
+                            field_defs,
+                        },
+                    );
+                    // enter self-definition for struct name
+                    use_def_map.insert(
+                        name_start.line,
+                        UseDef::new(
+                            references,
+                            pos.file_hash(),
+                            name_start,
+                            pos.file_hash(),
+                            name_start,
+                            name,
+                        ),
+                    );
+                }
+                None => {
+                    debug_assert!(false);
+                    continue;
+                }
+            };
+        }
+
+        for (pos, name, _) in &mod_def.constants {
+            match Self::get_start_loc(&pos, files, file_id_mapping) {
+                Some(name_start) => {
+                    constants.insert(*name, name_start);
+                    // enter self-definition for const name
+                    use_def_map.insert(
+                        name_start.line,
+                        UseDef::new(
+                            references,
+                            pos.file_hash(),
+                            name_start,
+                            pos.file_hash(),
+                            name_start,
+                            name,
+                        ),
+                    );
+                }
+                None => {
+                    debug_assert!(false);
+                    continue;
+                }
+            };
+        }
+
+        for (pos, name, _) in &mod_def.functions {
+            match Self::get_start_loc(&pos, files, file_id_mapping) {
+                Some(name_start) => {
+                    functions.insert(*name, name_start);
+                    // enter self-definition for const name
+                    use_def_map.insert(
+                        name_start.line,
+                        UseDef::new(
+                            references,
+                            pos.file_hash(),
+                            name_start,
+                            pos.file_hash(),
+                            name_start,
+                            name,
+                        ),
+                    );
+                }
+                None => {
+                    debug_assert!(false);
+                    continue;
+                }
+            };
+        }
+
+        let fhash = loc.file_hash();
+        (
+            ModuleDefs {
+                fhash,
+                structs,
+                constants,
+                functions,
+            },
```

Comment:
and a mega nit, I tend to not like it when my tuples span multiple lines. I think it just reads funny. But its a personal preference. 

I would do 
```
let module_defs =  ModuleDefs {
                fhash,
                structs,
                constants,
                functions,
            };
(module_defs,   use_def_map)
```
to avoid that (ignoring my horrible indentation)